api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java

```diff
@@ -425,85 +431,177 @@ public PatientIdentifierType getPatientIdentifierType(Integer patientIdentifierT
+						.getClassMetadata(PatientIdentifier.class));
+				if (aep != null) {
+
+					String[] properties = aep.getPropertyColumnNames(s);
+					if (properties != null && properties.length >= 1)
+						s = properties[0];
+				}
+
+				whereConditions.add(" t4." + s + " = t5." + s);
+				innerFields.add("pi1." + s);
+			} else {
+				log.warn("Unidentified attribute: " + s);
 			}
-			
-			if (!attributes.contains("includeVoided")) {
-				where += "and p1.voided = false and p2.voided = false ";
+		}
+
+		String innerFieldsJoined = StringUtils.join(innerFields, ", ");
+		String whereFieldsJoined = StringUtils.join(whereConditions, " and ");
+		String innerWhereCondition = "";
+		if (!attributes.contains("includeVoided")) {
+			innerWhereCondition = " where p1.voided = false ";
+		}
+		String innerQuery = "(Select " + innerFieldsJoined + innerSelect + innerWhereCondition + " group by "
+				+ innerFieldsJoined + " having count(*) > 1" + " order by " + innerFieldsJoined + ") t5";
+		return outerSelect + ", " + innerQuery + " where " + whereFieldsJoined + ";";
+	}
+
+	private void sortDuplicatePatients(List<Patient> patients, List<Integer> patientIds) {
+
+		Map<Integer, Integer> patientIdOrder = new HashMap<Integer, Integer>();
+		int startPos = 0;
+		for (Integer id : patientIds) {
+			patientIdOrder.put(id, startPos++);
+		}
+		class PatientIdComparator implements Comparator<Patient> {
+
+			private Map<Integer, Integer> sortOrder;
+
+			public PatientIdComparator(Map<Integer, Integer> sortOrder) {
+				this.sortOrder = sortOrder;
 			}
-			
-			for (String s : attributes) {
-				if (patientFieldNames.contains(s)) {
-					where += " and p1." + s + " = p2." + s;
-					orderBy += "p1." + s + ", ";
-				} else if (personFieldNames.contains(s)) {
-					if (!select.contains("Person ")) {
-						select += ", Person person1, Person person2";
-						where += " and p1.patientId = person1.personId and p2.patientId = person2.personId ";
-					}
-					where += " and person1." + s + " = person2." + s;
-					orderBy += "person1." + s + ", ";
-				} else if (personNameFieldNames.contains(s)) {
-					if (!select.contains("PersonName")) {
-						select += ", PersonName pn1, PersonName pn2";
-						where += " and p1 = pn1.person and p2 = pn2.person ";
-					}
-					where += " and pn1." + s + " = pn2." + s;
-					orderBy += "pn1." + s + ", ";
-				} else if (identifierFieldNames.contains(s)) {
-					if (!select.contains("PatientIdentifier")) {
-						select += ", PatientIdentifier pi1, PatientIdentifier pi2";
-						where += " and p1 = pi1.patient and p2 = pi2.patient ";
-					}
-					where += " and pi1." + s + " = pi2." + s;
-					orderBy += "pi1." + s + ", ";
-				} else {
-					log.warn("Unidentified attribute: " + s);
+
+			@Override
+			public int compare(Patient i1, Patient i2) {
+				Integer weekdayPos1 = sortOrder.get(i1.getPatientId());
+				if (weekdayPos1 == null) {
+					throw new IllegalArgumentException("Bad patient encountered: " + i1.getPatientId());
```

Comment:
Would it be faster if we dropped the order by clause in the first query? Since we end up having to sort in memory?